Upgrade to Electron v14.2.3

tidy-dev opened the pull request:
## Description

This PR bumps Electron to the latest version of v14. With the bump to 14, the electron team moved the `remote` module into it's own package. This upgrade requires install `@electron/remote` and updating all of the `remote` imports to import from that package. To use this `remote` package, we have to invoke the `initialize` method in the main processor and invoke the `enable` method on the window `webContents` (both app and crash windows). This "should" be a lift and shift and not impact functionality of any of the places we utilize remote, but I have tested those places since they are most likely regression areas.

UPDATE: Tho the build and testing it out was working. Unit test were failing `electron_1.ipcRenderer.sendSync` type errors. I am assuming there should be a way to just tell the unit tests that that type is ok.. but in the instances that those occurred, it could be resolved by removing the dependency on the `remote` module. That is something we want to do in all cases `remote` so I went ahead with that route.

Also, the setting of [`worldSafeExecuteJavaScript` was removed](https://www.electronjs.org/docs/latest/breaking-changes#removed-worldsafeexecutejavascript) and defaults to true - we do not use either of the methods that are specified as being impacted by this change.

Here is link to version[ 14 change log.](https://www.electronjs.org/releases/stable?version=14)

## Release notes
Notes: [Improved] Upgrade to Electron 14.2.3

tidy-dev committed:
Upgrade to 14.0.0 (build broken)

tidy-dev committed:
Update remote references to use new remote module

tidy-dev committed:
Initialize remote and enable it on the web contents

tidy-dev committed:
Upgrade to latest 14.2.3

tidy-dev committed:
Update zoom and window state on window load

tidy-dev committed:
Address themeChangeMonitor disposal unit test error

tidy-dev committed:
Auto updater unit test failure - updated per electron docs

Receiving unit test error: electron_1.ipcRenderer.sendSync

Revert private/public done during debugging

tidy-dev committed:
Remove remote dependency from change monitor

tidy-dev committed:
Use send.. cause window unit tests doesn't like invoke..

tidy-dev committed:
Remove remote dependency in windows controls

tidy-dev committed:
Lint...

sergiou87 commented on app/src/lib/stores/app-store.ts:
```diff
@@ -406,7 +402,7 @@ export class AppStore extends TypedBaseStore<IAppState> {
   private stashedFilesWidth = constrain(defaultStashedFilesWidth)
 
-  private windowState: WindowState
+  private windowState: WindowState = 'normal'
```
Should we initialize this to `null` or `undefined` instead? Checking the usages of this variable, I didn't see how setting 'normal' here could ever be a problem. However, I think using `null` or `undefined` would be more honest in case it actually matters.

sergiou87 commented on app/src/ui/window/window-controls.tsx:
```diff
@@ -34,7 +30,7 @@ interface IWindowControlState {
 export class WindowControls extends React.Component<{}, IWindowControlState> {
   public componentWillMount() {
-    this.setState({ windowState: getWindowState(remote.getCurrentWindow()) })
+    this.setState({ windowState: 'normal' })
```
Same about my other comment: maybe this should be initialized to `null` / `undefined`

sergiou87 commented on app/src/ui/lib/update-store.ts:
```diff
@@ -182,11 +189,13 @@ class UpdateStore {
-      autoUpdater.checkForUpdates()
-    } catch (e) {
-      this.emitError(e)
+    const checkedResults = await ipcRenderer.invoke(
+      'check-for-updates',
+      updatesURL
+    )
+
+    if (checkedResults !== null && checkedResults !== undefined) {
+      this.emitError(checkedResults)
     }
```
Probably worth renaming this to `error`?
```suggestion
    const error = await ipcRenderer.invoke(
      'check-for-updates',
      updatesURL
    )

    if (error !== null && error !== undefined) {
      this.emitError(error)
    }
```

sergiou87 commented on app/src/main-process/app-window.ts:
```diff
@@ -309,4 +328,66 @@ export class AppWindow {
+    autoUpdater.removeAllListeners()
+  }
+
+  public subscribeNativeThemUpdated() {
```
Typo here:
```suggestion
  public subscribeNativeThemeUpdated() {
```

tidy-dev committed:
Better variable name

Co-Authored-By: Sergio Padrino <1083228+sergiou87@users.noreply.github.com>

tidy-dev committed:
Fix typo

Co-Authored-By: Sergio Padrino <1083228+sergiou87@users.noreply.github.com>

tidy-dev committed:
Initialize window state to null

Co-Authored-By: Sergio Padrino <1083228+sergiou87@users.noreply.github.com>

tidy-dev committed:
Merge branch 'development' into electron-v14

niik commented on app/src/lib/stores/app-store.ts:
```diff
@@ -504,16 +500,12 @@ export class AppStore extends TypedBaseStore<IAppState> {
       error => this.emitError(error)
     )
 
-    const browserWindow = remote.getCurrentWindow()
-    this.windowState = getWindowState(browserWindow)
-
-    this.onWindowZoomFactorChanged(browserWindow.webContents.zoomFactor)
```
Why isn't this necessary any more? I would think we'd have to get a hold of the zoomFactor before it changes (like if the app starts with a zoomFactor other than 1)

tidy-dev replied:
Ditto to other comment.

niik commented on app/src/lib/stores/app-store.ts:
```diff
@@ -504,16 +500,12 @@ export class AppStore extends TypedBaseStore<IAppState> {
     )
 
-    const browserWindow = remote.getCurrentWindow()
-    this.windowState = getWindowState(browserWindow)
```
Wouldn't we want to initialize this (even if not synchronously?)

tidy-dev replied:
Seeing your other comment, I guess you saw that I am initializing this as I was sending these values on app-window load. Tho I didn't know the concern about that possibly loading before the app-store since it tested out well.  But, no matter, working to use the invoke/handle to request the values from here (2nd stab at that.. was running into issue the first time I tried that).

niik commented on app/src/main-process/app-window.ts:
```diff
@@ -145,6 +159,11 @@ export class AppWindow {
 
     this.window.webContents.on('did-finish-load', () => {
       this.window.webContents.setVisualZoomLevelLimits(1, 1)
+      this.window.webContents.send(
+        'zoom-factor-changed',
+        this.window.webContents.zoomFactor
+      )
+      sendWindowStateEvent(this.window, getWindowState(this.window))
```
I don't think this is safe as we can't know for certain that the AppStore is set up to receive these events when `did-finish-load` happens, can we? `did-finish-load` is an Electron event that signals that the page has loaded. `renderer-ready` would be a better approach but I think the safest approach would be to request these states from the AppStore as we did before, only asynchronously this time round.

tidy-dev replied:
I tried the approach of going from the app-store first, but I had other conflicting issue at the time.. so going to try again. Thanks for explaining the timing of 'did-finish-load' - I did test it out and the values were being set as I expected on the 'did-finish-load' but it sounds like that may not be an guaranteed case.

niik reviewed (changes requested):
Thanks for tackling this best of an upgrade @tidy-dev! I had a few questions.

In general I'd like for us to try to minimize `ipcRenderer.invoke(` calls all over the codebase and would prefer it if followed the existing convention of proxy functions in `main-process-proxy.ts`. Now that we're rapidly expanding the number of IPC calls and events we might also want to contemplate sharing the IPC event name and possibly even the type signature between the main process and the renderer so that we don't end up introducing bugs due to a typo in the event name.

tidy-dev committed:
Use ipcRenderer to initialize window state and zoom

tidy-dev committed:
Add invoke to ipcRenderer mock

tidy-dev committed:
Use main process proxy for auto updater methods

tidy-dev committed:
Refactor theme change monitor to use main process proxy

tidy-dev committed:
Refactor windows controls to use main proxy process

say25 commented:
Don’t forget to update this file as well https://github.com/desktop/desktop/blob/main/app/.npmrc

tidy-dev committed:
Add another jest mock

tidy-dev committed:
Update .npmrc

tidy-dev committed:
Remove test that will be removed with 16 upgrade

tidy-dev committed:
Don't do integration test - there are not any

tidy-dev committed:
Remove it from the yml..

tidy-dev committed:
Remove spectron since unused and rest of integration test references

tidy-dev committed:
With spectron gone - need to add lodash typing.

tidy-dev committed:
Merge branch 'development' into electron-v14

tidy-dev commented:
There is one unit test that I was having trouble getting this branch to pass -> it is the one we usually see - the flaky one. But I think it was exasperated with a typing error now (which adding to the mock file didn't appear to help).  When I looked into it, it is the one we want to remove when we upgrade to 16 anyways because we are going to remove the spectron dependency. Thus, I went ahead and removed `spectron` and removed the integration test file (and related references - removed integration test call from ci.yml). Evidently `spectron` was bringing us `lodash` typing so I had to add that in separately.

tidy-dev committed:
Merge branch 'development' into electron-v14

tidy-dev committed:
Update the 14 upgraded added ipc calls to use strongly typed ipc calls

tidy-dev committed:
Use new ipcWebContents

tidy-dev committed:
Put wrapper methods back with note to why they are important

niik commented on app/src/ui/main-process-proxy.ts:
```diff
@@ -46,6 +46,79 @@ export const showFolderContents = sendProxy('show-folder-contents')
+  'subscribe-native-theme-updated'
+)
+
+/** Subscribes to the native theme updated ecent originating from the main process */
```
```suggestion
/** Subscribes to the native theme updated event originating from the main process */
```

niik commented on app/src/main-process/app-window.ts:
```diff
@@ -307,4 +319,86 @@ export class AppWindow {
+    nativeTheme.addListener('updated', (event: string, userInfo: any) => {
+      ipcWebContents.send(this.window.webContents, 'native-theme-updated')
+    })
+  }
```
This pattern lends itself to excessive subscriptions if we find another place in the renderer where we'd like to get notified of theme changes. I think we should either always subscribe to the event in the main process and send it along to the renderer such that the subscription in the renderer becomes the only deciding factor. Alternatively we can keep the current approach but make the method idempotent

```ts
  private subscribed = false
  
  public subscribeNativeThemeUpdated() {
    if (!this.subscribed) {
      nativeTheme.addListener('updated', this.onNativeThemeUpdated)
      this.subscribed = true
    }
  }

  private onNativeThemeUpdated = () =>
    ipcWebContents.send(this.window.webContents, 'native-theme-updated')
```

tidy-dev replied:
I like the subscribe in main process - less messaging code to follow.

niik commented on app/src/main-process/app-window.ts:
```diff
@@ -74,6 +85,7 @@ export class AppWindow {
     })
 
     ipcMain.on('will-quit', event => {
+      nativeTheme.removeAllListeners()
```
Given that the `native-theme-updated` event is scoped to the main window only I think it would make more sense to have `mainWindow` handle the unsubscription. The `will-quit` was an acceptable approach when using the renderer as we could rely on the remote module unsubscribing ipc events when the process exited.

```ts
this.window.on('closed', () => {
  nativeTheme.removeListener('updated', this.onNativeThemeUpdated)
})
```

```suggestion
```

niik commented on app/src/ui/lib/update-store.ts:
```diff
@@ -55,25 +62,20 @@ class UpdateStore {
+
+      window.addEventListener('beforeunload', () => {
+        disposeAutoUpdater()
+      })
```
This should be the responsibility of the main process, `beforeunload` won't trigger if we destroy the window from the main process (which we do in the event of an uncaught exception for example).

Similar to the `nativeTheme` unloading above we should handle this in the `closed` event on the app window.

tidy-dev replied:
Also, subscribed to the auto updater on load like the native theme.

niik commented on app/src/ui/lib/update-store.ts:
```diff
@@ -55,25 +62,20 @@ class UpdateStore {
+      window.addEventListener('beforeunload', () => {
+        disposeAutoUpdater()
+      })
+    } catch (e) {
+      this.emitError(e)
+    }
```
What kind of exception are we expecting here? It seems like wiring up event listeners should be fairly safe right?

tidy-dev replied:
I don't know - That was copied from original renderer code. I didn't dig into why it was. Happy to remove it - especially since the auto updates appears to capture it errors and send them as event.

niik commented on app/src/main-process/main.ts:
```diff
@@ -431,6 +433,50 @@ app.on('ready', () => {
+
+  ipcMain.on('close-window', () => {
+    mainWindow?.closeWindow()
+  })
```
We can tighten all of these (and several of the other as well) up by using the short hand arrow function syntax
```suggestion
  ipcMain.on('minimize-window', () => mainWindow?.minimizeWindow())
  ipcMain.on('maximize-window', () => mainWindow?.maximizeWindow())
  ipcMain.on('unmaximize-window', () => mainWindow?.unmaximizeWindow())
  ipcMain.on('close-window', () => mainWindow?.closeWindow())
```

niik commented on package.json:
```diff
@@ -8,12 +8,11 @@
     "cli": "ts-node --require ./app/test/globals.ts --require ./app/src/cli/dev-commands-global.ts app/src/cli/main.ts",
     "check:eslint": "tsc -P eslint-rules/",
     "test:eslint": "jest eslint-rules/tests/*.test.js",
-    "test:integration": "cross-env TEST_ENV=1 ELECTRON_NO_ATTACH_CONSOLE=1 xvfb-maybe --auto-servernum -- jest --testLocationInResults --config ./app/jest.integration.config.js",
```
I think we can remove the `xvfb-maybe` dependency now as well, right?

tidy-dev replied:
Yep. Good catch!

niik committed:
Don't crash in prod if we fail to send an ipc message

sergiou87 reviewed (dismissed):
Thanks for making those changes! LGTM 😄

tidy-dev committed:
Typo

Co-Authored-By: Markus Olsson <634063+niik@users.noreply.github.com>

tidy-dev committed:
One theme change monitor subscription

Co-Authored-By: Markus Olsson <634063+niik@users.noreply.github.com>

tidy-dev committed:
Subscribe/Dispose of auto updater events in main process

Co-Authored-By: Markus Olsson <634063+niik@users.noreply.github.com>

tidy-dev committed:
Remove unnecessary try/catch block

Co-Authored-By: Markus Olsson <634063+niik@users.noreply.github.com>

tidy-dev committed:
Tightening up ipc function calls

Co-Authored-By: Markus Olsson <634063+niik@users.noreply.github.com>

tidy-dev committed:
Remove unused dependency xvfb-maybe

Co-Authored-By: Markus Olsson <634063+niik@users.noreply.github.com>

tidy-dev committed:
Merge branch 'development' into electron-v14

tidy-dev committed:
Merge branch 'development' into electron-v14

tidy-dev committed:
Merge branch 'development' into electron-v14

niik commented on app/src/lib/window-state.ts:
```diff
@@ -56,7 +56,7 @@ export function registerWindowStateChangedEvents(
  * over the window-state-changed channel to the render process.
  */
-function sendWindowStateEvent(
+export function sendWindowStateEvent(
```
Did we need this export?

```suggestion
function sendWindowStateEvent(
```

niik reviewed (dismissed):
One tiny question but feel free to ignore it and merge anyway. Great work on this @tidy-dev!

tidy-dev committed:
Remove unneeded export identifier

Co-authored-by: Markus Olsson <j.markus.olsson@gmail.com>
